--- backfire/package/brcm2708-gpu-fw/Makefile
+++ backfire/package/brcm2708-gpu-fw/Makefile
@@ -8,13 +8,13 @@
 include $(INCLUDE_DIR)/kernel.mk
 
 PKG_NAME:=brcm2708-gpu-fw
-PKG_REV:=d5b05be2147bf5dc0137798837af24b0bbbe398d
-PKG_VERSION:=20130517
+PKG_REV:=e25efc4545d0cef7c49c82e770b3e4ccfc8aba4c
+PKG_VERSION:=20140226
 PKG_RELEASE:=1
 
 PKG_SOURCE:=$(PKG_REV).tar.gz
 PKG_SOURCE_URL:=https://github.com/Hexxeh/rpi-firmware/archive/
-PKG_MD5SUM:=9b0ce0a530e237f4c6fe43a36ccf57c3
+PKG_MD5SUM:=b1bb6b1063c4f920314274edd39c35e6
 
 PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)/rpi-firmware-$(PKG_REV)
 
